File: TypeScript/Codewars/src/resolutions/Find-Nearest-square-number.ts
export function nearestSq(n: number): number {
	const squareRoot = Math.sqrt(n);

	if (Number.isInteger(squareRoot)) return n;

	const smallest = Math.floor(squareRoot) ** 2;

	const biggest = Math.ceil(squareRoot) ** 2;

	return nearestToX([smallest, biggest], n);
}

function nearestToX(values: number[], x: number): number {
	return values.reduce((prev, curr) => (Math.abs(curr - x) < Math.abs(prev - x) ? curr : prev));
}

File: Csharp/CodeWars/src/Completed/FindNearestSquareNumber.cs
namespace CodeWars.Completed;

public class FindNearestSquareNumber
{
    public static int NearestSq(int n)
    {
        double sqrt = Math.Sqrt(n);

        if (sqrt % 1 == 0)
            return n;

        int lower = (int)Math.Pow(Math.Floor(sqrt), 2);
        int upper = (int)Math.Pow(Math.Ceiling(sqrt), 2);

        return Math.Abs(n - lower) <= Math.Abs(n - upper) ? lower : upper;
    }
}
